tests/integration/test_exporter.py

```diff
@@ -0,0 +1,147 @@
+        ops_test.model.deploy(
+            APPLICATION_APP_NAME,
+            application_name=APPLICATION_APP_NAME,
+            num_units=1,
+            # MySQL Router and Grafana agent are subordinate -
+            # they will use the series of the principal charm
+            series=mysql_router_charm_series,
+            channel="latest/edge",
+        ),
+        ops_test.model.deploy(
+            GRAFANA_AGENT_APP_NAME,
+            application_name=GRAFANA_AGENT_APP_NAME,
+            num_units=0,
+            channel="latest/stable",
+            series=mysql_router_charm_series,
+        ),
+    )
+
+    mysql_app, mysql_router_app, mysql_test_app, grafana_agent_app = applications
+
+    logger.info("Relating mysqlrouter and grafana-agent with mysql-test-app")
+
+    await ops_test.model.relate(
+        f"{MYSQL_ROUTER_APP_NAME}:database", f"{APPLICATION_APP_NAME}:database"
+    )
+
+    await ops_test.model.relate(
+        f"{APPLICATION_APP_NAME}:juju-info", f"{GRAFANA_AGENT_APP_NAME}:juju-info"
+    )
+
+    async with ops_test.fast_forward():
+        await asyncio.gather(
+            ops_test.model.block_until(lambda: mysql_app.status == "active", timeout=SLOW_TIMEOUT),
+            ops_test.model.block_until(
+                lambda: mysql_router_app.status == "blocked", timeout=SLOW_TIMEOUT
+            ),
+            ops_test.model.block_until(
+                lambda: mysql_test_app.status == "waiting", timeout=SLOW_TIMEOUT
+            ),
+            ops_test.model.block_until(
+                lambda: grafana_agent_app.status == "blocked", timeout=SLOW_TIMEOUT
+            ),
+        )
+
+        logger.info("Relating mysqlrouter with mysql")
+
+        await ops_test.model.relate(
+            f"{MYSQL_ROUTER_APP_NAME}:backend-database", f"{MYSQL_APP_NAME}:database"
+        )
+
+        await asyncio.gather(
+            ops_test.model.block_until(lambda: mysql_app.status == "active", timeout=SLOW_TIMEOUT),
+            ops_test.model.block_until(
+                lambda: mysql_router_app.status == "active", timeout=SLOW_TIMEOUT
+            ),
+            ops_test.model.block_until(
+                lambda: mysql_test_app.status == "active", timeout=SLOW_TIMEOUT
+            ),
+            ops_test.model.block_until(
+                lambda: grafana_agent_app.status == "blocked", timeout=SLOW_TIMEOUT
+            ),
+        )
+
+    unit = mysql_test_app.units[0]
+    unit_address = await unit.get_public_address()
+
+    try:
+        http.request("GET", f"http://{unit_address}:49152/metrics")
+        assert False, "❌ can connect to metrics endpoint without relation with cos"
+    except urllib3.exceptions.MaxRetryError as e:
+        assert (
+            "[Errno 111] Connection refused" in e.reason.args[0]
+        ), "❌ expected connection refused error"
+
+    await ops_test.model.relate(
+        f"{GRAFANA_AGENT_APP_NAME}:cos-agent", f"{MYSQL_ROUTER_APP_NAME}:cos-agent"
+    )
+
+    time.sleep(30)
```

Comment:
since the condition we would wait for is internal to the charm (a service needs to become available within the charm), i prefer to use sleep in this test. we are checking that an endpoint for that service returns a response after a fixed amount of time (30s)